src/algebra/algebra/unitization.lean

```diff
@@ -0,0 +1,505 @@
+  [smul_comm_class R A A] : comm_monoid (unitization R A) :=
+{ mul_comm := λ x₁ x₂, ext (mul_comm x₁.1 x₂.1) $
+    show x₁.1 • x₂.2 + x₂.1 • x₁.2 + x₁.2 * x₂.2 = x₂.1 • x₁.2 + x₁.1 • x₂.2 + x₂.2 * x₁.2,
+    by rw [add_comm (x₁.1 • x₂.2), mul_comm],
+  ..unitization.monoid }
+
+instance [comm_semiring R] [non_unital_semiring A] [module R A] [is_scalar_tower R A A]
+  [smul_comm_class R A A] : semiring (unitization R A) :=
+{ ..unitization.monoid,
+  ..unitization.non_assoc_semiring }
+
+-- This should work for `non_unital_comm_semiring`s, but we don't seem to have those
+instance [comm_semiring R] [comm_semiring A] [module R A] [is_scalar_tower R A A]
+  [smul_comm_class R A A] : comm_semiring (unitization R A) :=
+{ ..unitization.comm_monoid,
+  ..unitization.non_assoc_semiring }
+
+variables (R A)
+
+/-- The canonical inclusion of rings `R →+* unitization R A`. -/
+@[simps apply]
+def inl_ring_hom [semiring R] [non_unital_semiring A] [module R A] : R →+* unitization R A :=
+{ to_fun := inl,
+  map_one' := inl_one A,
+  map_mul' := inl_mul A,
+  map_zero' := inl_zero A,
+  map_add' := inl_add A }
+
+end mul
+
+section algebra
+variables (S R A : Type*)
+[comm_semiring S] [comm_semiring R] [non_unital_semiring A]
+[module R A] [is_scalar_tower R A A] [smul_comm_class R A A]
+[algebra S R] [distrib_mul_action S A] [is_scalar_tower S R A]
+
+instance algebra : algebra S (unitization R A) :=
+{ commutes' := λ r x,
+  begin
+    induction x using unitization.ind,
+    simp only [mul_add, add_mul, ring_hom.to_fun_eq_coe, ring_hom.coe_comp, function.comp_app,
+      inl_ring_hom_apply, inl_mul_inl],
+    rw [inl_mul_inr, inr_mul_inl, mul_comm]
+  end,
+  smul_def' := λ s x,
+  begin
+    induction x using unitization.ind,
+    simp only [mul_add, smul_add, ring_hom.to_fun_eq_coe, ring_hom.coe_comp, function.comp_app,
+      inl_ring_hom_apply, algebra.algebra_map_eq_smul_one],
+    rw [inl_mul_inl, inl_mul_inr, smul_one_mul, inl_smul, inr_smul, smul_one_smul]
+  end,
+  ..(unitization.inl_ring_hom R A).comp (algebra_map S R) }
+
+lemma algebra_map_eq_inl_comp : ⇑(algebra_map S (unitization R A)) = inl ∘ algebra_map S R := rfl
+lemma algebra_map_eq_inl_ring_hom_comp :
+  algebra_map S (unitization R A) = (inl_ring_hom R A).comp (algebra_map S R) := rfl
+lemma algebra_map_eq_inl : ⇑(algebra_map R (unitization R A)) = inl := rfl
+lemma algebra_map_eq_inl_hom : algebra_map R (unitization R A) = inl_ring_hom R A := rfl
+
+/-- The canonical `R`-algebra projection `unitization R A → R`. -/
+@[simps]
+def fst_hom : unitization R A →ₐ[R] R :=
+{ to_fun := fst,
+  map_one' := fst_one,
+  map_mul' := fst_mul,
+  map_zero' := fst_zero,
+  map_add' := fst_add,
+  commutes' := fst_inl A }
+
+end algebra
+
+section coe
+variables {R A : Type*}
+
+instance [has_zero R] : has_coe_t A (unitization R A) := { coe := inr }
+
+@[simp]
+lemma coe_apply [has_zero R] (a : A) : (a : unitization R A) = inr a := rfl
```

Comment:
I'm not certain whether this coercion is a good idea. We don't have a coercion from the generators of the free group into the free group etc, but I guess we do have it for `with_one`.

At any rate, we shouldn't have both `coe` and `inr` as valid spellings - if you much prefer `coe`, then lets just use that everywhere in this file and never define `inr`.

Reply:
I think I would prefer `coe` for sure. We really want to think of the non-unital algebra as canonically sitting inside its unitization. The (minimal) unitization has nice universal properties. I'll switch `inr` to `coe` throughout.